#' @title Computes XP-EHH Statistics (standardized cross-population EHH)
#'
#' @description
#' This function computes XP-EHH Statistics (standardized cross-population EHH) using two populations
#' and target-SNPs, then performs the SNPs' annotation and generate appropriate tables and graphics.
#'
#' @usage xpehh(pop1, pop2, popname1, popname2, snp.list = "all", filter = 2, method = "both",
#' annot = T, write.xls = "both", plot = T)
#'
#' @param pop1 Scalar character with the name of the first population to be compared to the second.
#' @param pop2 Scalar character with the name of the second population to be compared to the first one.
#' @param popname1 Scalar character with an optional name for the first population.
#' @param popname2 Scalar character with an optional name for the second population.
#' @param snp.list Character vector. It could be a scalar named "all" (default) for loading all SNPs
#' present in 'snps' directory, or a character vector with the exact names from the target-SNP files
#' (also supposed to be in 'populations' directory).
#' @param filter Numeric value for filtering significant XP-EHH scores (default = 2).
#' @param method Scalar character. Indicates if applied method must be "bilateral", "unilateral",
#' or "both" (default).
#' @param annot Logical scalar. If TRUE (default), SNP annotation in xls files will be written with
#' correspondent SNP's genes, as well as ancestral and derived allele present in the dataset.
#' FALSE turns off this feature.
#' @param write.xls Scalar character. It could be set to "all.snps" for annotating all SNPs included
#' in this analysis, and "ss.snps" for annotating only SNPs statistically significant
#' (defined by 'filter' argument), or 'both' (default) for both annotation outputs.
#'
#'
#' @param plot Logical scalar. If TRUE (default), then XP-EHH-related plots are generated and saved
#' as png figures.
#'
#'
#' @details
#' R binary file named 'scanhh.RData' must be loaded before running this function. Populations going
#' to be analyzed must be provided as 'scanhh_list$NAME', where 'NAME' is the population name (see @@Examples
#' section). This function creates a new directory in 'rehh_out' folder, named 'xpehh', with two sub-directories:
#' 'xpehh/graphics' for the plots, and 'ihs/tables' for csv and xls files.
#'
#' The 'method' parameter indicates which method will be applied ('bilateral', 'unilateral', or 'both';
#' default = 'both'). The default value for 'filter' parameter has been set to 2 accordingly the
#' standard value described on original paper.
#'
#' The 'annot' parameter turns on or off annotation procedure (default is TRUE, meaning 'on'), while argument
#' passed in 'write.xls' parameter tells which SNPs must be included on analysis ("all.snps", "ss.snps"
#' or "both"). Plot argument enables the generation of two plots/figures per population.
#'
#' If all parameter's arguments are used as they are by default, then a total of four simple files (.csv)
#' containing XP-EHH results will be generated by analysis to both methods and both 'write.xls' arguments.
#'
#' Except by annotation argument ('annot = T'), this function usually runs very fast. However,
#' depending on how many target-SNPs are present in the dataset, the annotation velocity may vary considerably.
#'
#' @return XP-EHH files: simple result files (.csv), SNP annotation (.xls) and/or iHS plots (.png).
#' @seealso \code{\link{make.scanhh}}
#'
#' @examples
#' \dontrun{
#' load('scanhh.RData')
#' xpehh(pop1 = scanhh_list$Africans, pop2 = scanhh_list$Europeans)}
#'
#' \dontrun{
#' load('scanhh.RData')
#' xpehh(pop1 = scanhh_list$Africans, pop2 = scanhh_list$Europeans), snp.list = 'all', filter = 2,
#' method = 'unilateral', annot = T, write.xls = 'ss.snps', plot = T)
#' }
#'
#' @export
#'

xpehh <- function(pop1, pop2, popname1, popname2, snp.list = "all", filter = 2, method = "both", annot = T, write.xls = "both",
    plot = T) {

    # Computing XP-EHH Statistics as described by Sabeti et al. XXXX
    load("scanhh.RData")

    if (missing(pop1)) {
        stop("You must specify the populations in 'scanhh.list' file.\nExample:\n> load(scanhh.list)\n> pop1 = scanhh_list$population1")
    }

    if (missing(pop2)) {
        stop("You must specify the populations in 'scanhh.list' file.\nExample:\n> load(scanhh.list)\n> pop2 = scanhh_list$population2")
    }

    if (missing(popname1)) {
        popname1 <- comment(pop1)
    }

    if (missing(popname2)) {
        popname2 <- comment(pop2)
    }


    if (snp.list[1] == "all") {
        snp.list <- list.files(path = "snps/", all.files = TRUE)
        snp.list <- grep(pattern = "[[:alnum:]]", x = snp.list, value = TRUE)
        snp.list.data <- unlist(lapply(X = snp.list, FUN = function(x) readLines(paste0("snps/", x))))
    } else {
        snp.list.data <- unlist(lapply(X = snp.list, FUN = function(x) readLines(paste0("snps/", x))))
    }

    cat("\n Computing XP-EHH Statistics... \n")
    dir.create(path = "rehh_out/xpehh", showWarnings = F)
    dir.create(path = "rehh_out/xpehh/tables", showWarnings = F)

    if (method == "bilateral" | method == "both") {

        xpehh.bi <- rehh::ies2xpehh(pop1, pop2, popname1, popname2, method = "bilateral")
        xpehh.bi.snp.list <- xpehh.bi[row.names(xpehh.bi) %in% snp.list.data, ]
        xpehh.bi.snp.list <- xpehh.bi.snp.list[complete.cases(xpehh.bi.snp.list), ]
        xpehh.bi.snp.list.stat <- xpehh.bi.snp.list[xpehh.bi.snp.list[, 3] >= filter | xpehh.bi.snp.list[, 3] <= -filter,
            ]

        write.csv(xpehh.bi.snp.list, file = paste0("./rehh_out/xpehh/tables/", popname1, ".vs.", popname2, ".xpehh.bi.csv"),
            quote = F, row.names = T)
        write.csv(xpehh.bi.snp.list.stat, file = paste0("./rehh_out/xpehh/tables/", popname1, ".vs.", popname2, ".xpehh.bi.stat.csv"),
            quote = F, row.names = T)
    }

    if (method == "unilateral" | method == "both") {

        xpehh.uni <- rehh::ies2xpehh(pop1, pop2, popname1, popname2, method = "unilateral")
        xpehh.uni.snp.list <- xpehh.uni[row.names(xpehh.uni) %in% snp.list.data, ]
        xpehh.uni.snp.list <- xpehh.uni.snp.list[complete.cases(xpehh.uni.snp.list), ]
        xpehh.uni.snp.list.stat <- xpehh.uni.snp.list[xpehh.uni.snp.list[, 3] >= filter | xpehh.uni.snp.list[, 3] <= -filter,
            ]

        write.csv(xpehh.uni.snp.list, file = paste0("./rehh_out/xpehh/tables/", popname1, ".vs.", popname2, ".xpehh.uni.csv"),
            quote = F, row.names = T)
        write.csv(xpehh.uni.snp.list.stat, file = paste0("./rehh_out/xpehh/tables/", popname1, ".vs.", popname2, ".xpehh.uni.stat.csv"),
            quote = F, row.names = T)
    }

    ### Generating graphics ###

    if (plot == TRUE) {
        cat(" Plotting XPEHH results... \n\n")
        dir.create(path = "rehh_out/xpehh/graphics", showWarnings = FALSE)

        xpehh.list <- list(xpehh.bi, xpehh.uni)
        names(xpehh.list) <- c("xpehh.bi", "xpehh.uni")

        par(mfrow = c(1, 1))
        rehh::xpehhplot(xpehh.list$xpehh.bi, main = names(xpehh.bi.snp.list)[3])
        dev.set(dev.prev())
        savePlot(filename = paste0("rehh_out/xpehh/graphics/", popname1, ".vs.", popname2, ".xpehhplot.bi.score.png"),
            type = "png")
        dev.off()
        savePlot(filename = paste0("rehh_out/xpehh/graphics/", popname1, ".vs.", popname2, ".xpehhplot.bi.log.png"), type = "png")
        dev.off()

        par(mfrow = c(1, 1))
        rehh::xpehhplot(xpehh.list$xpehh.uni, main = names(xpehh.bi.snp.list)[3])
        dev.set(dev.prev())
        savePlot(filename = paste0("rehh_out/xpehh/graphics/", popname1, ".vs.", popname2, ".xpehhplot.uni.score.png"),
            type = "png")
        dev.off()
        savePlot(filename = paste0("rehh_out/xpehh/graphics/", popname1, ".vs.", popname2, ".xpehhplot.uni.log.png"),
            type = "png")
        dev.off()
    }

    # SNP Annotation devtools::install_github('cran/NCBI2R') library(NCBI2R)

    if (annot == T) {
        load("haplotypes/haps-sample.RData")
        haps.info.alleles <- do.call(rbind, haps.info)


        if (method == "bilateral" | method == "both") {

            if (write.xls == "all.snps" | write.xls == "both") {
                tryCatch({
                  alleles.state <- haps.info.alleles[haps.info.alleles$V2 %in% rownames(xpehh.bi.snp.list), ]
                  xpehh.bi.snp.list.ncbi <- NCBI2R::GetSNPInfo(rownames(xpehh.bi.snp.list), pbar = F, showurl = F)
                  xpehh.bi.snp.list.ncbi <- xpehh.bi.snp.list.ncbi$genesymbol
                  xpehh.bi.snp.list.ncbi.xls <- cbind(CHR = xpehh.bi.snp.list$CHR, SNP = rownames(xpehh.bi.snp.list),
                    Allele_A = alleles.state$V4, Allele_D = alleles.state$V5, GENE = xpehh.bi.snp.list.ncbi, xpehh.bi.snp.list[2:4])
                }, error = function(e) {
                  cat("\tNo SNPs have been found in this dataset (method = \"bilateral\" & write.xls = \"all.snps\")\n")
                })

                if (exists("xpehh.bi.snp.list.ncbi.xls")) {
                  file.name <- paste0("rehh_out/xpehh/tables/", popname1, ".vs.", popname2, ".xpehh.bi.all.xls")
                  WriteXLS::WriteXLS(xpehh.bi.snp.list.ncbi.xls, ExcelFileName = file.name, SheetNames = "XP-EHH Analysis",
                    AdjWidth = T, BoldHeaderRow = T)
                  cat(paste0(paste0("File '", popname1, ".vs.", popname2, ".xpehh.bi.all.xls"), "' successfully saved into 'rehh_out/ihs/tables/' directory \n"))
                } else {
                  message(paste0("\t", nrow(xpehh.bi.snp.list), " SNP(s) remained after XP-EHH analysis for the SNPs analyzed\n"))
                }
            }

            if (write.xls == "ss.snps" | write.xls == "both") {
                tryCatch({
                  alleles.state <- haps.info.alleles[haps.info.alleles$V2 %in% rownames(xpehh.bi.snp.list.stat), ]
                  xpehh.bi.snp.list.stat.ncbi <- NCBI2R::GetSNPInfo(rownames(xpehh.bi.snp.list.stat), pbar = F)
                  xpehh.bi.snp.list.stat.ncbi <- xpehh.bi.snp.list.stat.ncbi$genesymbol
                  xpehh.bi.snp.list.ncbi.xls <- cbind(CHR = xpehh.bi.snp.list.stat$CHR, SNP = rownames(xpehh.bi.snp.list.stat),
                    Allele_A = alleles.state$V4, Allele_D = alleles.state$V5, GENE = xpehh.bi.snp.list.stat.ncbi, xpehh.bi.snp.list.stat[2:4])
                }, error = function(e) {
                  cat("\tNo statistical significant SNPs have been found in this dataset (method = \"bilateral\" & write.xls = \"ss.snps\")\n")
                })

                if (exists("xpehh.bi.snp.list.ncbi.xls")) {
                  file.name <- paste0("rehh_out/xpehh/tables/", popname1, ".vs.", popname2, ".xpehh.bi.stat.xls")
                  WriteXLS::WriteXLS(xpehh.bi.snp.list.ncbi.xls, ExcelFileName = file.name, SheetNames = "XP-EHH Analysis",
                    AdjWidth = T, BoldHeaderRow = T)
                  cat(paste0(paste0("File '", popname1, ".vs.", popname2, ".xpehh.bi.stat.xls"), "' successfully saved into 'rehh_out/ihs/tables/' directory \n"))
                } else {
                  message(paste0("\t", nrow(xpehh.bi.snp.list.stat), " SNP(s) remained after XP-EHH analysis for the SNPs analyzed\n"))
                }
            }
        }

        if (method == "unilateral" | method == "both") {

            if (write.xls == "all.snps" | write.xls == "both") {
                tryCatch({
                  alleles.state <- haps.info.alleles[haps.info.alleles$V2 %in% rownames(xpehh.uni.snp.list), ]
                  xpehh.uni.snp.list.ncbi <- NCBI2R::GetSNPInfo(rownames(xpehh.uni.snp.list), pbar = F)
                  xpehh.uni.snp.list.ncbi <- xpehh.uni.snp.list.ncbi$genesymbol
                  xpehh.uni.snp.list.ncbi.xls <- cbind(CHR = xpehh.uni.snp.list$CHR, SNP = rownames(xpehh.uni.snp.list),
                    Allele_A = alleles.state$V4, Allele_D = alleles.state$V5, GENE = xpehh.uni.snp.list.ncbi, xpehh.uni.snp.list[2:4])
                }, error = function(e) {
                  cat("\tNo SNPs have been found in this dataset (method = \"unilateral\" & write.xls = \"all.snps\")\n")
                })

                if (exists("xpehh.uni.snp.list.ncbi.xls")) {
                  file.name <- paste0("rehh_out/xpehh/tables/", popname1, ".vs.", popname2, ".xpehh.uni.all.xls")
                  WriteXLS::WriteXLS(xpehh.uni.snp.list.ncbi.xls, ExcelFileName = file.name, SheetNames = "XP-EHH Analysis",
                    AdjWidth = T, BoldHeaderRow = T)
                  cat(paste0(paste0("File '", popname1, ".vs.", popname2, ".xpehh.uni.all.xls"), "' successfully saved into 'rehh_out/ihs/tables/' directory \n"))
                } else {
                  message(paste0("\t", nrow(xpehh.uni.snp.list), " SNP(s) remained after XP-EHH analysis for the SNPs analyzed\n"))
                }
            }

            if (write.xls == "ss.snps" | write.xls == "both") {

                alleles.state <- haps.info.alleles[haps.info.alleles$V2 %in% rownames(xpehh.uni.snp.list.stat), ]
                tryCatch({
                  xpehh.uni.snp.list.stat.ncbi <- NCBI2R::GetSNPInfo(rownames(xpehh.uni.snp.list.stat), pbar = F)
                  xpehh.uni.snp.list.stat.ncbi <- xpehh.uni.snp.list.stat.ncbi$genesymbol
                  xpehh.uni.snp.list.stat.ncbi.xls <- cbind(CHR = xpehh.uni.snp.list.stat$CHR, SNP = rownames(xpehh.uni.snp.list.stat),
                    Allele_A = alleles.state$V4, Allele_D = alleles.state$V5, GENE = xpehh.uni.snp.list.stat.ncbi, xpehh.uni.snp.list.stat[2:4])
                }, error = function(e) {
                  cat("\tNo statistical significant SNPs have been found in this dataset (method = \"unilateral\" & write.xls = \"ss.snps\")\n")
                })

                if (exists("xpehh.uni.snp.list.stat.ncbi.xls")) {
                  file.name <- paste0("rehh_out/xpehh/tables/", popname1, ".vs.", popname2, ".xpehh.uni.stat.xls")
                  WriteXLS::WriteXLS(xpehh.uni.snp.list.stat.ncbi.xls, ExcelFileName = file.name, SheetNames = "XP-EHH Analysis",
                    AdjWidth = T, BoldHeaderRow = T)
                  cat(paste0(paste0("File '", popname1, ".vs.", popname2, ".xpehh.uni.stat.xls"), "' successfully saved into 'rehh_out/ihs/tables/' directory \n"))
                } else {
                  message(paste0("\t", nrow(xpehh.uni.snp.list.stat), " SNP(s) remained after XP-EHH analysis for the SNPs analyzed\n"))
                }
            }
        }
    }
}
